File: src/Solutions/Euler353.hs
module Solutions.Euler353 where
import Control.Monad
import Control.Monad.ST
import Data.Array
import Data.Array.ST
import Data.List
import Text.Printf
import Utils.Array

type Coords = (Double, Double, Double)

triples :: Int -> [Coords]
triples r = [ (fromIntegral a, fromIntegral b, fromIntegral c)
            | a <- [0..r], b <- [0..r]
            , let c2 = r * r - a * a - b * b
            , c2 >= 0
            , let c = floor $ sqrt $ fromIntegral c2
            , c * c == c2 ]

-- https://stackoverflow.com/questions/52210911/great-circle-distance-between-two-p-x-y-z-points-on-a-unit-sphere
risk :: Coords -> Coords -> Double
risk (a, b, c) (x, y, z) = (2 * phi / pi)^2
    where rawDist = sqrt $ (a - x)^2 + (b - y)^2 + (c - z)^2
          phi = asin $ rawDist / (2 * (sqrt $ a^2 + b^2 + c^2))
reflectRisk :: Coords -> Double
reflectRisk (a, b, c) = risk (a, b, c) (a, b, -c)

infty = 10**10 :: Double

-- Assumption: Road is symmetric
dijkstra :: [Coords] -> Double
dijkstra coords = runST $ do
    dist <- newArray (1, l) infty :: ST s (STUArray s Int Double)
    visited <- newArray (1, l) False :: ST s (STUArray s Int Bool) 
    writeArray dist 1 0
    let findMinIndex = forM [1..l] (\x -> readArray visited x >>= \p -> if p
                                    then return (x, infty + 1)
                                    else readArray dist x >>= \v -> return (x, v)) >>=
                       return . fst . foldr (\x@(_, dMin) y@(_, d) -> if d < dMin then y else x) (0, infty)
    let loop = findMinIndex >>= \i -> when (i > 0) $ do
                  writeArray visited i True
                  thisDist <- readArray dist i
                  forM_ [1..l] $ \j -> when (i /= j) $
                      modifyArray dist j $ min $ thisDist + risk (verticesArr ! i) (verticesArr ! j)
                  loop
    loop
    fmap (foldr min infty) $ forM [1..l] $ \i -> do
        v <- readArray dist i
        return $ 2 * v + reflectRisk (verticesArr ! i)
    where vertices = zip [1..] coords
          l = length coords
          verticesArr = array (1, l) vertices

funM :: Int -> Double
funM = dijkstra . triples

euler353 :: IO String
euler353 = return $ printf "%.10f" $ sum $ map (funM . (\x -> 2^x - 1)) [1..15]

File: rust/src/solutions/euler353.rs
use num::integer::Roots;
use ordered_float::OrderedFloat;
use std::f64::{consts::PI, INFINITY};

#[derive(Copy, Clone)]
struct Coords {
    x: f64,
    y: f64,
    z: f64,
}

fn triples(r: i64) -> Vec<Coords> {
    let mut ans = vec![];
    for a in 0..=r {
        for b in 0..=r {
            let c2 = r * r - a * a - b * b;
            if c2 >= 0 {
                let c = c2.sqrt();
                if c * c == c2 {
                    ans.push(Coords {
                        x: a as f64,
                        y: b as f64,
                        z: c as f64,
                    });
                }
            }
        }
    }
    ans
}

fn risk(a: Coords, b: Coords) -> f64 {
    let raw_dist = ((a.x - b.x).powi(2) + (a.y - b.y).powi(2) + (a.z - b.z).powi(2)).sqrt();
    let phi = (raw_dist / 2.0 / (a.x.powi(2) + a.y.powi(2) + a.z.powi(2)).sqrt()).asin();
    (2.0 * phi / PI).powi(2)
}

fn reflect_risk(a: Coords) -> f64 {
    let mut b = a;
    b.z = -b.z;
    risk(a, b)
}

fn dijkstra(coords: Vec<Coords>) -> f64 {
    let l = coords.len();
    let mut dist = vec![OrderedFloat(INFINITY); l];
    let mut visited = vec![false; l];
    dist[0] = OrderedFloat(0.0);
    while let Some(i) = (0..l).filter(|&i| !visited[i]).min_by_key(|&i| dist[i]) {
        visited[i] = true;
        for j in 0..l {
            if i != j {
                dist[j] = std::cmp::min(
                    dist[j],
                    OrderedFloat(dist[i].0 + risk(coords[i], coords[j])),
                );
            }
        }
    }
    (0..l)
        .map(|i| 2.0 * dist[i].0 + reflect_risk(coords[i]))
        .min_by_key(|&p| OrderedFloat(p))
        .unwrap()
}

pub fn euler353() -> String {
    let ans = std::iter::successors(Some(2), |&x| Some(x + x))
        .take(15)
        .map(|x| dijkstra(triples(x - 1)))
        .sum::<f64>();
    format!("{:.10}", ans)
}
